spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/ComputeTableStatsSparkAction.java

```diff
@@ -0,0 +1,150 @@
+import org.apache.iceberg.types.Types;
+import org.apache.spark.sql.SparkSession;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Computes the statistics of the given columns and stores it as Puffin files. */
+public class ComputeTableStatsSparkAction extends BaseSparkAction<ComputeTableStatsSparkAction>
+    implements ComputeTableStats {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ComputeTableStatsSparkAction.class);
+
+  private final Table table;
+  private Set<String> columns;
+  private Snapshot snapshot;
+
+  ComputeTableStatsSparkAction(SparkSession spark, Table table) {
+    super(spark);
+    this.table = table;
+    this.snapshot = table.currentSnapshot();
+    this.columns =
+        table.schema().columns().stream().map(Types.NestedField::name).collect(Collectors.toSet());
+  }
+
+  @Override
+  protected ComputeTableStatsSparkAction self() {
+    return this;
+  }
+
+  @Override
+  public Result execute() {
+    String desc = String.format("Computing stats for %s", table.name());
+    JobGroupInfo info = newJobGroupInfo("COMPUTE-TABLE-STATS", desc);
+    return withJobGroupInfo(info, this::doExecute);
+  }
+
+  private Result doExecute() {
+    if (snapshot == null) {
+      return ImmutableComputeTableStats.Result.builder().build();
+    }
+    LOG.info("Computing stats of {} for snapshot {}", table.name(), snapshot.snapshotId());
+    List<Blob> blobs = generateNDVBlobs();
+    StatisticsFile statisticFile;
+    try {
+      statisticFile = writeAndCommitPuffin(blobs);
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    }
+    return ImmutableComputeTableStats.Result.builder().statisticsFile(statisticFile).build();
+  }
+
+  private StatisticsFile writeAndCommitPuffin(List<Blob> blobs) throws IOException {
+    LOG.info(
+        "Writing stats to puffin files for table {} for snapshot{}",
+        table.name(),
+        snapshot.snapshotId());
+    TableOperations operations = ((HasTableOperations) table).operations();
+    String path = operations.metadataFileLocation(String.format("%s.stats", UUID.randomUUID()));
+    OutputFile outputFile = operations.io().newOutputFile(path);
+    GenericStatisticsFile statisticsFile = generateStatisticsFile(blobs, outputFile, path);
+    table.updateStatistics().setStatistics(snapshot.snapshotId(), statisticsFile).commit();
+    return statisticsFile;
+  }
+
+  @NotNull
+  private GenericStatisticsFile generateStatisticsFile(
+      List<Blob> blobs, OutputFile outputFile, String path) throws IOException {
+    try (PuffinWriter writer =
+        Puffin.write(outputFile).createdBy("Iceberg ComputeTableStats action").build()) {
+      blobs.forEach(writer::add);
+      writer.finish();
+      return new GenericStatisticsFile(
+          snapshot.snapshotId(),
+          path,
+          writer.fileSize(),
+          writer.footerSize(),
+          writer.writtenBlobsMetadata().stream()
```

Comment:
Optional: What about overloading `from` in `GenericStatisticsFile` to accept a `Collection` of blobs?

Reply:
Nope, I meant something like this to `GenericBlobMetadata`. We have `from` there that accepts a single object so let's overload it to accept a list/collection of objects.

```
public static List<BlobMetadata> from(
    Collection<org.apache.iceberg.puffin.BlobMetadata> puffinMetadataList) {
  return puffinMetadataList.stream()
      .map(GenericBlobMetadata::from)
      .collect(ImmutableList.toImmutableList());
}
```